### PR TITLE
Error handling for calls (and room)

### PR DESCRIPTION
**Pull request by bwindels**
This includes the error boundary from #978, see `docs/error-handling.md` for a description of how and when the error handling is tackled here. I've added this error handling to the calls code, and also to the room view model as I needed it to report errors when starting a call.

### CODE REVIEW

File: src/domain/session/room/CallViewModel.ts [hunk 1]
```diff
@@ -15,11 +15,13 @@ limitations under the License.
...
 
 import {AvatarSource} from "../../AvatarSource";
-import {ViewModel, Options as BaseOptions} from "../../ViewModel";
+import type  {ViewModel} from "../../ViewModel";
+import {ErrorReportViewModel, Options as BaseOptions} from "../../ErrorReportViewModel";
 import {getStreamVideoTrack, getStreamAudioTrack} from "../../../matrix/calls/common";
 import {avatarInitials, getIdentifierColorNumber, getAvatarHttpUrl} from "../../avatar";
 import {EventObservableValue} from "../../../observable/value/EventObservableValue";
 import {ObservableValueMap} from "../../../observable/map/ObservableValueMap";
+import { ErrorViewModel } from "../../ErrorViewModel";
```
Comment (MidhunSureshR) [inline src/domain/session/room/CallViewModel.ts]:
```suggestion
import {ErrorViewModel} from "../../ErrorViewModel";
```

File: src/domain/session/room/CallViewModel.ts [hunk 2]
```diff
@@ -28,22 +30,32 @@ import type {BaseObservableList} from "../../../observable/list/BaseObservableLi
 import type {BaseObservableValue} from "../../../observable/value/BaseObservableValue";
 import type {Stream} from "../../../platform/types/MediaDevices";
 import type {MediaRepository} from "../../../matrix/net/MediaRepository";
+import type { Session } from "../../../matrix/Session";
```
Comment (MidhunSureshR) [inline src/domain/session/room/CallViewModel.ts]:
```suggestion
import type {Session} from "../../../matrix/Session";
```

File: src/domain/rageshake.ts [hunk 3]
```diff
@@ -63,3 +67,28 @@ export async function submitLogsToRageshakeServer(data: RageshakeData, logsBlob:
...
+            app: "hydrogen",
+            userAgent: platform.description,
+            version: platform.version,
+            text: `Submit logs from settings for user ${session.userId} on device ${session.deviceId}`,
+        },
+        logExport.asBlob(),
+        bugReportEndpointUrl,
+        platform.request
+    );
+}
```
Comment (MidhunSureshR) [inline src/domain/rageshake.ts]: We should add a newline here and at the end of other files.

File: src/domain/session/room/RoomViewModel.js [hunk 4]
```diff
@@ -20,14 +20,16 @@ import {ComposerViewModel} from "./ComposerViewModel.js"
 import {CallViewModel} from "./CallViewModel"
 import {PickMapObservableValue} from "../../../observable/value/PickMapObservableValue";
 import {avatarInitials, getIdentifierColorNumber, getAvatarHttpUrl} from "../../avatar";
+import {ErrorReportViewModel} from "../../ErrorReportViewModel";
 import {ViewModel} from "../../ViewModel";
 import {imageToInfo} from "../common.js";
 import {LocalMedia} from "../../../matrix/calls/LocalMedia";
+import { ErrorViewModel } from "../../ErrorViewModel";
```
Comment (MidhunSureshR) [inline src/domain/session/room/RoomViewModel.js]:
```suggestion
import {ErrorViewModel} from "../../ErrorViewModel";
```

File: src/domain/session/room/RoomViewModel.js [hunk 5]
```diff
@@ -143,14 +141,8 @@ export class RoomViewModel extends ViewModel {
...
-        if (this._timelineError) {
-            return `Something went wrong loading the timeline: ${this._timelineError.message}`;
-        }
-        if (this._sendError) {
-            return `Something went wrong sending your message: ${this._sendError.message}`;
-        }
-        return "";
+    get errorViewModel() {
+        return this._errorViewModel;
     }
```
Comment (MidhunSureshR) [inline src/domain/session/room/RoomViewModel.js]: Do we not already inherit this getter from `ErrorReportViewModel`?

File: src/domain/session/room/RoomViewModel.js [hunk 6]
```diff
@@ -289,10 +278,8 @@ export class RoomViewModel extends ViewModel {
             attachments["info.thumbnail_url"] = 
                 this._room.createAttachment(thumbnail.blob, file.name);
             await this._room.sendEvent("m.room.message", content, attachments);
-        } catch (err) {
-            this._sendError = err;
-            this.emitChange("error");
-            console.error(err.stack);
+        } catch (error) {
+            this.reportError(error);
```
Comment (MidhunSureshR) [inline src/domain/session/room/RoomViewModel.js]: So the way I understand it, `reportError()` will eventually lead to a UI showing the error and giving the user the option to submit logs. But I don't see where the error passed to `reportError()` is logged?

**Review by MidhunSureshR — COMMENTED** [review]
Looks good, just need to point out some formatting issues :+1:

File: src/domain/session/room/RoomViewModel.js (hunk 5 above)
Comment (bwindels) [reply src/domain/session/room/RoomViewModel.js]: We do! good catch

File: src/domain/session/room/RoomViewModel.js (hunk 6 above)
Comment (bwindels) [reply src/domain/session/room/RoomViewModel.js]: Indeed, that is how it works. You mean apart from here? I've made a convenience method, `ErrorReportViewModel.logAndCatch`, to combine logging and error handling in one. Perhaps that's why you see `reportError` being called less than expected?

**Review by MidhunSureshR — APPROVED** [review]
Lgtm!

**Pull request merged**